### PR TITLE
Default the label for CCB when reading cached labels

### LINKED ISSUE #2116: Using multi-pass learning with CCB causes a seg fault
Repro: `vw test/train-sets/ccb_test.dat --ccb_explore_adf --passes 2 -c`

```
/mnt/c/w/repos/vw_linux$ vw test/train-sets/ccb_test.dat --ccb_explore_adf --passes 2 -c
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
using cache_file = test/train-sets/ccb_test.dat.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.000000 0.000000            1            1.0 0:0.000000,1:0.000000,... 0,1,2,...       18
0.000000 0.000000            2            2.0 3:0.000000,4:0.000000,... 3,4,0,...       18
-0.250000 -0.500000            4            4.0 1:0.000000,4:-1.000000,... 1,4,2,...       18
-0.542586 -0.835171            8            8.0 4:-1.000000,1:0.000000,... 4,1,0,...       18
-0.327906 -0.041667           16           16.0  ?,?,... 4,2,1,...       18
Segmentation fault (core dumped)
```

### LINKED ISSUE #2170: CCB Memory Leak
Two memory leaks:

Test 201:
```
==198== error calling PR_SET_PTRACER, vgdb might block
==198== 64 (40 direct, 24 indirect) bytes in 1 blocks are definitely lost in loss record 2 of 2
==198==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==198==    by 0x43FD26: CCB::parse_outcome(substring&) (ccb_label.cc:231)
==198==    by 0x440BF3: CCB::parse_label(parser*, shared_data*, void*, v_array<substring>&) (ccb_label.cc:315)
==198==    by 0x1AB056: substring_to_example(vw*, example*, substring) (parse_example.cc:461)
==198==    by 0x1AADC1: read_features_string(vw*, v_array<example*>&) (parse_example.cc:45)
==198==    by 0x12CB2A: parse_dispatch(vw&, std::function<void (vw&, v_array<example*>&)>) (parse_dispatch_loop.h:18)
==198==    by 0x1BCC0F: main_parse_loop(vw*) (parser.cc:906)
==198==    by 0x1C0A3A: void std::__invoke_impl<void, void (*)(vw*), vw*>(std::__invoke_other, void (*&&)(vw*), vw*&&) (invoke.h:60)
==198==    by 0x1BF85B: std::__invoke_result<void (*)(vw*), vw*>::type std::__invoke<void (*)(vw*), vw*>(void (*&&)(vw*), vw*&&) (invoke.h:95)
==198==    by 0x1D2A04: decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)())) std::thread::_Invoker<std::tuple<void (*)(vw*), vw*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (thread:234)
==198==    by 0x1D1583: std::thread::_Invoker<std::tuple<void (*)(vw*), vw*> >::operator()() (thread:243)
==198==    by 0x1D02CB: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(vw*), vw*> > >::_M_run() (thread:186)
==198==
```

Test 202:
```
==221== error calling PR_SET_PTRACER, vgdb might block
==221== 12 bytes in 1 blocks are definitely lost in loss record 1 of 5
==221==    at 0x4C2FA3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==221==    by 0x4C31D84: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==221==    by 0x1D20D0: v_array<unsigned int>::resize(unsigned long) (v_array.h:74)
==221==    by 0x1D0B6F: v_array<unsigned int>::push_back(unsigned int const&) (v_array.h:112)
==221==    by 0x1CCB2D: LabelObjectState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:254)
==221==    by 0x1CD866: DefaultState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:825)
==221==    by 0x1CA105: VWReaderHandler<false>::EndObject(unsigned int) (parse_example_json.h:1322)
==221==    by 0x1C7C1D: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:637)
==221==    by 0x1C4602: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1398)
==221==    by 0x1C7F0B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseArray<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:674)
==221==    by 0x1C461B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1399)
==221==    by 0x1C7B50: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:621)
==221== 
==221== 36 bytes in 1 blocks are definitely lost in loss record 3 of 5
==221==    at 0x4C31D2F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==221==    by 0x1D20D0: v_array<unsigned int>::resize(unsigned long) (v_array.h:74)
==221==    by 0x1D0B6F: v_array<unsigned int>::push_back(unsigned int const&) (v_array.h:112)
==221==    by 0x1CCB2D: LabelObjectState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:254)
==221==    by 0x1CD866: DefaultState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:825)
==221==    by 0x1CA105: VWReaderHandler<false>::EndObject(unsigned int) (parse_example_json.h:1322)
==221==    by 0x1C7C1D: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:637)
==221==    by 0x1C4602: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1398)
==221==    by 0x1C7F0B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseArray<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:674)
==221==    by 0x1C461B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1399)
==221==    by 0x1C7B50: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:621)
==221==    by 0x1C4602: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1398)
==221== 
==221== 176 (80 direct, 96 indirect) bytes in 2 blocks are definitely lost in loss record 5 of 5
==221==    at 0x4C3017F: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==221==    by 0x1CCB9F: LabelObjectState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:260)
==221==    by 0x1CD866: DefaultState<false>::EndObject(Context<false>&, unsigned int) (parse_example_json.h:825)
==221==    by 0x1CA105: VWReaderHandler<false>::EndObject(unsigned int) (parse_example_json.h:1322)
==221==    by 0x1C7C1D: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:637)
==221==    by 0x1C4602: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1398)
==221==    by 0x1C7F0B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseArray<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:674)
==221==    by 0x1C461B: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1399)
==221==    by 0x1C7B50: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseObject<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:621)
==221==    by 0x1C4602: void rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::ParseValue<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:1398)
==221==    by 0x1C3286: rapidjson::ParseResult rapidjson::GenericReader<rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator>::Parse<1u, rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >, VWReaderHandler<false> >(rapidjson::GenericInsituStringStream<rapidjson::UTF8<char> >&, VWReaderHandler<false>&) (reader.h:501)
==221==    by 0x1C1AA1: void VW::read_line_decision_service_json<false>(vw&, v_array<example*>&, char*, unsigned long, bool, example& (*)(void*), void*, DecisionServiceInteraction*) (parse_example_json.h:1405)
==221== 
```

### PR DESCRIPTION
**Pull request by jackgerrits**
Fixes #2116
Fixes #2170

This fix is a bit concerning. I don't know why cached labels are not defaulted before being written into.

### CODE REVIEW

File: test/unit_test/ccb_parser_test.cc [hunk 1]
```diff
@@ -24,93 +24,106 @@ BOOST_AUTO_TEST_CASE(ccb_parse_label)
...
-	  CCB::label label;
-	  parse_label(lp, &p, "ccb shared", label);
-	  BOOST_CHECK_EQUAL(label.explicit_included_actions.size(), 0);
-	  BOOST_CHECK(label.outcome == nullptr);
-	  BOOST_CHECK_EQUAL(label.type, CCB::example_type::shared);
-	  lp.delete_label(&label);
+    CCB::label label;
+    // Memset is used because in practice these labels are always Calloced.
+    // This is dangerous though and we should REALLY let this object have a constructor.
+    memset(&label, 0, sizeof(label));
```
Comment (lokitoth) [inline test/unit_test/ccb_parser_test.cc]: Can you use the scoped calloc mechanism that we use in VW for this? In other words:

```
auto ptr = scoped_calloc...()
CCB::label& label = *ptr;
```

File: vowpalwabbit/ccb_label.cc [hunk 2]
```diff
@@ -161,8 +165,16 @@ void cache_label(void* v, io_buf& cache)
 void default_label(void* v)
 {
   CCB::label* ld = static_cast<CCB::label*>(v);
-  ld->outcome = nullptr;
-  ld->explicit_included_actions = v_init<uint32_t>();
+
+  // This is tested against nullptr, so unfortunately as things are this must be deleted when not used.
+  if(ld->outcome)
```
Comment (lokitoth) [inline vowpalwabbit/ccb_label.cc]: We should probably start tagging places where better encapsulation with RAII would be nice (but is pending on figuring out how to deal with unions going forward)

Comment (jackgerrits) [reply vowpalwabbit/ccb_label.cc]: I agree
